File: CPP/110.BalancedBinaryTree.cpp
#include <bits/stdc++.h>
using namespace std;
/*
    bool isBalanced(TreeNode* root) {
        if(root==nullptr) return true;
        return abs(height(root->left)-height(root->right))<=1 && isBalanced(root->left) && isBalanced(root->right);
    }
    int height(TreeNode*root){
        if(root==nullptr) return 0;
        return max(height(root->left),height(root->right))+1;
    }
*/

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

map<TreeNode*,pair<int,int>> m;

//calculate maxHeight (left and right) for all nodes
int calcHeight(TreeNode* root) {
    if (root==nullptr) return -1;
    int lH=1+calcHeight(root->left);
    int rH=1+calcHeight(root->right);
    pair<int,int> p1=make_pair(lH,rH);
    //cout << p1.first << " " << p1.second << " " << endl;
    m[root]=p1;
    return max(lH,rH);
}

void preOrderVisit(TreeNode* root) {
    if (root==nullptr) return;
    cout << root->val << " ";
    preOrderVisit(root->left);
    preOrderVisit(root->right);
}

bool isBalanced(TreeNode* root) {
    m.clear();
    calcHeight(root);
    int maxDif=0;
    for (auto entry: m) {
        auto el=entry.second;
        //cout << el.first << " " << el.second << endl;
        maxDif=max(maxDif,abs(el.first-el.second));
    }
    return maxDif<=1;
}

int main() {
    TreeNode *bt1 = new TreeNode(3,new TreeNode(9),new TreeNode(20,new TreeNode(15),new TreeNode(7)));
    //preOrderVisit(bt1);
    cout << isBalanced(bt1) << endl; //true

    TreeNode *bt2 = new TreeNode(1,new TreeNode(2,new TreeNode(3,new TreeNode(4),new TreeNode(4)),new TreeNode(3)),new TreeNode(2));
    //preOrderVisit(bt2);
    cout << isBalanced(bt2) << endl; //false

    TreeNode *bt3 = new TreeNode();
    //preOrderVisit(bt2);
    cout << isBalanced(bt3) << endl; //false

    return 0;
}

File: Python/110.BalancedBinaryTree.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root or (not root.left and not root.right): return True
        isBal=True
        if root.left and root.right: isBal=-1<=self.maxDepth(root.left)-self.maxDepth(root.right)<=1
        if root.left: 
            isBal=isBal and self.isBalanced(root.left)
            if not root.right:
                if self.maxDepth(root.left)>1: return False
        if root.right: 
            isBal=isBal and self.isBalanced(root.right)
            if not root.left:
                if self.maxDepth(root.right)>1: return False
        return isBal

    def maxDepth(self, root):
        if not root: return 0
        return 1+max(self.maxDepth(root.right),self.maxDepth(root.left))

s=Solution()
t1=TreeNode(3,TreeNode(9),TreeNode(20,TreeNode(15),TreeNode(7)))
# print(s.isBalanced(t1)) # True

t2=TreeNode(1,None,TreeNode(2,None,TreeNode(3)))
# print(s.isBalanced(t2)) # False

t3=TreeNode(1,TreeNode(2),None)
print(s.isBalanced(t3)) # True